proposals/0436-recordwildcards-total.rst

```diff
@@ -0,0 +1,125 @@
+Proposal title
+==============
+
+.. author:: Brandon Chinn
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/436>`_.
+.. contents::
+
+``-XRecordWildCards`` does not give any warning or indication if you forget to use a field. This proposal adds syntax (optionally enabled by another extension) to give a warning if a field is unused.
+
+
+Motivation
+----------
+
+Some functions require all the fields of a record to be used, or at least considered, e.g. ``hash`` or ``toJSON``, or even something more security-facing. Currently, if one uses ``RecordWildCards`` and they add a field to the record later on, nothing will tell the developer to take a look at these functions.
+
+::
+
+ data Foo = Foo { a :: Int, b :: Bool }
+ 
+ instance ToJSON Foo where
+   toJSON Foo{..} =
+     object
+       [ "a" .= a
+       , "b" .= b
+       ]
+       
+ instance Hashable Foo where
+   hash Foo{..} = hash (a, b)
+ 
+ -- adding a 'c' field does not raise any indication of needing to look at `toJSON` or `hash`
+
+Existing workarounds are described in `this blog post <https://cs-syd.eu/posts/2021-09-10-undefined-trick>`_, all of which leave something to be desired:
+
+* Positional arguments break when fields are reordered, which matters in functions like `toJSON`
+* Using `undefined` is a bit of a hack, plus it fails `-Wunused-pattern-binds`. It also fails to explicitly say which field is unused (which isn't strictly necessary, but it's nice to have)
+
+The `safe-wild-cards <https://hackage.haskell.org/package/safe-wild-cards>`_ package provides a Template Haskell solution to this, but it would be nice to support this natively.
+
+One could even think of this as the dual of the `Or pattern proposal <https://github.com/ghc-proposals/ghc-proposals/pull/43>`_, to warn on unused terms in product types as opposed to unused terms in sum types.
+
+Proposed Change Specification
+-----------------------------
+
+When ``-XRecordWildCards`` is enabled, in addition to supporting ``Foo{..}`` in a pattern match, also support ``Foo{..!}``, which desugars to ``Foo{a = a, b = b}``.
```

Comment:
I agree with the motivation for this proposal (came here to check whether such a proposal existed), disagree with the proposed change.

Currently the `!` token in Haskell means that `seq` is involved, but the difference between `Foo{..}` and `Foo{..!}` is not a presence / absence of `seq`. So at the very least, I think the proposed `!` should be ruled out. More generally, though, I don't like using a new syntactic form to trigger the checks we need.

Here are some alternatives:

* Change [`-Wunused-record-wildcards`](https://downloads.haskell.org/ghc/latest/docs/users_guide/using-warnings.html#ghc-flag-Wunused-record-wildcards)

  Currently the flag warns when *none* of the fields in the pattern `Foo{..}` are used. This was the first flag I tried when I wanted completeness warnings for records, and I was surprised by its behaviour. I expected it to tell warn me for each field that was unused.

* Add `-Wunused-record-wildcard-fields`

  A backwards-compatible version of the previous point. While `-Wunused-record-wildcards` tells you that an entire `Foo{..}` is redundant, `-Wunused-record-wildcard-fields` will tell you each field that is unused.

* Add `-Wincomplete-record-patterns`

  I don't *need* a RecordWildCards-based solution. `-Wincomplete-record-patterns` will warn when a record pattern doesn't name all the fields associated with the constructor. Using the datatype from the proposal, `Foo{a = x}` would generate a "field `b` unused" warning, and so would `Foo{a}` under NamedFieldPuns. If RecordWildCards can actually be specified in terms of NamedFieldPuns (i.e. `Foo{..}` desugars to `Foo{a, b}` desugars to `Foo{a = a, b = b}`), then `-Wincomplete-record-patterns` should also give the warnings described in the previous two points.